crates/ursa-gateway/src/cache/tlrfu.rs

```diff
@@ -0,0 +1,468 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+    time::UNIX_EPOCH,
+};
+
+use anyhow::{bail, Context, Result};
+
+use super::lru::Lru;
+use crate::util::timer::_now;
+
+struct Data {
+    value: Vec<u8>,
+    freq: usize,
+    lru_k: usize,
+    _tll: u128,
+}
+
+pub struct Tlrfu {
+    store: HashMap<Arc<String>, Data>,
+    freq: BTreeMap<usize, Lru<usize, Arc<String>>>, // shrinkable
```

Comment:
so the outermost map is for us to track the "frequency" with a counter and the inner map's (LRU) key is to track the least recently used?

Reply:
Exactly 💯